GNUmakefile

```diff
@@ -0,0 +1,76 @@
+# User-customizable variables:
+CPPFLAGS ?=
+CXX ?= c++
+CXXFLAGS ?=
+DEPFLAGS ?= -MD -MF $(@).d -MP -MT $(@)
+LDFLAGS ?=
+LDLIBS ?=
+LOADLIBES ?=
+OUTPUT_OPTION ?= -o $(@)
+TARGET_ARCH ?=
+build_dir = build
+
+.DELETE_ON_ERROR:
+.SECONDARY:
+.SUFFIXES:
+
+example_programs = cli_ws_deque
+test_programs = ntest/ntest
+
+example_exe_files = $(foreach name,$(example_programs),$(build_dir)/example/$(name)/$(name))
+test_exe_files = $(foreach name,$(test_programs),$(build_dir)/test/$(name))
+
+exe_files = $(example_exe_files) $(test_exe_files)
+o_files = $(exe_files:=.cpp.o)
+
+ansi_term_csi = [
+ansi_term_bold = $(ansi_term_csi)1m
+ansi_term_green = $(ansi_term_csi)32m
+ansi_term_red = $(ansi_term_csi)31m
+ansi_term_reset = $(ansi_term_csi)m
+
+COMPILE.cpp = $(CXX) $(DEPFLAGS) $(CXXFLAGS) $(CPPFLAGS) $(TARGET_ARCH) -c
+LINK.cpp = $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(TARGET_ARCH)
+
+.PHONY: all
+all: examples tests
+
+.PHONY: examples tests
+examples: $(example_exe_files)
+tests: $(test_exe_files)
+
+.PHONY: check
+check: check-examples check-tests
+
+.PHONY: check-examples check-tests
+check-examples: $(example_exe_files:=.check-result)
+check-tests: $(test_exe_files:=.check-result)
+
+$(build_dir)/%.check-result: $(build_dir)/% always-run
```

Comment:
Can't we declare this is `.PHONY` instead of using the fake `always-run` dependency? How is it different?

Reply:
> Can't we declare this is `.PHONY`

I wish! That was the first thing I tried. Frustratingly, `.PHONY` doesn't work in this case.

### Proposal A: `.PHONY`

With the following code, the `.PHONY` line accomplishes nothing, because `%` isn't handled:

```
.PHONY: $(build_dir)/%.check-result
$(build_dir)/%.check-result: $(build_dir)/%
```

Demonstration:

```
$ make
$ make check
example/cli_ws_deque/cli_ws_deque ...
ws_deque_test
iterations: 1000
[snip]
throughput: 100000

test/ntest/ntest OK

$ make check
make: Nothing to be done for `check'.
```

### Proposal B: `.PHONY` with explicit target list

In the following code, I took proposal A and manually expanded `%` in `.PHONY`.

```
.PHONY: $(build_dir)/example/cli_ws_deque/cli_ws_deque.check-result $(build_dir)/test/ntest/ntest.check-result
$(build_dir)/%.check-result: $(build_dir)/%
```

The effect is that the rules are ignored when running `make check`, and `make check` always does nothing:

```
$ make clean
$ make check
make: Nothing to be done for `check'.
$ make
$ make check
make: Nothing to be done for `check'.
```

See the following remark on [`.PHONY` and implicit rules](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html#index-phony-targets) from GNU `make`'s documentation:

> The implicit rule search (see [Implicit Rules](https://www.gnu.org/software/make/manual/html_node/Implicit-Rules.html#Implicit-Rules)) is skipped for `.PHONY` targets. This is why declaring a target as `.PHONY` is good for performance, even if you are not worried about the actual file existing.

Reply:
I see.

We could also remove `printf %d "$${status}" >$(@); \` line so that output files are not actually create. But I don't know what is better. Not producing output files looks a bit dirty.